kernel/libc/koslib/gmon.c

```diff
@@ -0,0 +1,386 @@
+
+    if(cxt->state != GMON_PROF_ON)
+        return;
+
+    /* If function is within the .text section */
+    if(frompc >= cxt->lowpc && frompc <= cxt->highpc) {
+        cxt->pc = selfpc; /* Save the last place we are */
+
+        /* Calculate index into 'froms' array */
+        index = (frompc - cxt->lowpc) / HASHFRACTION;
+        index_ptr = &cxt->froms[index];
+
+        /* Grab index into node array */
+	    node_index = *index_ptr;
+
+        /* Node doesnt exist? */
+        if(node_index == 0) {
+            goto create_node;
+        }
+
+        /* Try and find the node */
+        node = &cxt->nodes[node_index];
+
+        while(true) {
+            /* Found the node */
+            if(node->selfpc == selfpc) {
+                node->count++;
+                return;
+            }
+            /* Check if selfpc is less */
+            else if(selfpc < node->selfpc) {
+                if(node->left == 0) {
+                    index_ptr = &node->left;
+                    goto create_node;
+                }
+
+                node = &cxt->nodes[node->left];
+            }
+            else {
+                if(node->right == 0) {
+                    index_ptr = &node->right;
+                    goto create_node;
+                }
+
+                node = &cxt->nodes[node->right];
+            }
+        }
+
+create_node:
+        node_index = ++cxt->node_count;
+        if(node_index >= cxt->nnodes)
+            /* halt further profiling */
+            goto overflow;
+
+        *index_ptr = node_index;
+        node = &cxt->nodes[node_index];
+        node->selfpc = selfpc;
+        node->count = 1;
+        node->left = 0;
+        node->right = 0;
+        return;
+
+overflow:
+        cxt->state = GMON_PROF_ERROR;
+        //ERR("mcount: call graph buffer size limit exceeded, gmon.out will not be generated\n");
+        return;
+    }
+}
+
+void _monstartup(uintptr_t lowpc, uintptr_t highpc) {
+    size_t counter_size = 0;
+    size_t froms_size = 0;
+    size_t nodes_size = 0;
+    gmon_context_t *cxt = &g_context;
+
+    /* Exit early if we already initialized */
+    if(cxt->initialized)
+        return;
+
+    cxt->lowpc = ROUNDDOWN(lowpc, HISTFRACTION * sizeof(HIST_COUNTER_TYPE));
+    cxt->highpc = ROUNDUP(highpc, HISTFRACTION * sizeof(HIST_COUNTER_TYPE));
+    cxt->textsize = cxt->highpc - cxt->lowpc;
+
+    /* Calculate the number of counters, rounding up to ensure no remainder,
+       ensuring that all of textsize is covered without any part being left out */
+    cxt->ncounters = (cxt->textsize + HISTFRACTION - 1) / HISTFRACTION;
+    counter_size = cxt->ncounters * sizeof(HIST_COUNTER_TYPE);
+
+    cxt->nfroms = (cxt->textsize + HASHFRACTION - 1) / HASHFRACTION;
+    froms_size = cxt->nfroms * sizeof(uint16_t);
+
+    cxt->nnodes = (cxt->textsize * ARCDENSITY) / 100;
+    nodes_size = cxt->nnodes * sizeof(gmon_node_t);
+
+    printf("Low: %p, High: %p\n", (void *)cxt->lowpc, (void *)cxt->highpc);
```

Comment:
```suggestion
    printf("Low: %p, High: %p\n", (void *)cxt->lowpc, (void *)cxt->highpc);
```
I would probably change these to like `dbglog(DBG_VERBOSE, ...);` so that people like Shirobon don't bitch that KOS is too chatty... I would actually make the first log line something like `dbglog(DBG_NOTICE, "Enabling gprof from <%p to %p>\n", ...);` or something...

I'm of the opinion that we should be logging more *major events* such as starting to profile, a controller being added/removed, etc as `DBG_NOTICE` with subsequent more detailed output/information as `DBG_VERBOSE` or `DBG_DEBUG` so that we can capture useful logging info at different levels of granularity.

Reply:
My initial plan was to remove these after I was done debugging so that is why formatting is booboo. I will look into formatting them nicer and wrapping them in dbglog.